--- a/net/minecraft/world/entity/projectile/EntityShulkerBullet.java
+++ b/net/minecraft/world/entity/projectile/EntityShulkerBullet.java
@@ -59,8 +59,21 @@
     @Override
     public SoundCategory getSoundSource() {
         return SoundCategory.HOSTILE;
@@ -224,7 +237,7 @@
             MovingObjectPosition movingobjectposition = ProjectileHelper.getHitResult(this, this::canHitEntity);
 
             if (movingobjectposition.getType() != MovingObjectPosition.EnumMovingObjectType.MISS) {
-                this.onHit(movingobjectposition);
+                this.preOnHit(movingobjectposition); // CraftBukkit - projectile hit event
             }
         }
@@ -317,6 +330,11 @@
 
     @Override
     public boolean hurt(DamageSource damagesource, float f) {
         if (!this.level.isClientSide) {
             this.playSound(SoundEffects.SHULKER_BULLET_HURT, 1.0F, 1.0F);
             ((WorldServer) this.level).sendParticles(Particles.CRIT, this.getX(), this.getY(), this.getZ(), 15, 0.2D, 0.2D, 0.2D, 0.0D);
